pkg/virt-handler/ns/ns.go

```diff
@@ -0,0 +1,226 @@
+	Set() error
+
+	// Returns the filesystem path representing this object's namespace
+	Path() string
+
+	// Returns a file descriptor representing this object's namespace
+	Fd() uintptr
+
+	// Cleans up this instance of the namespace; if this instance
+	// is the last user the namespace will be destroyed
+	Close() error
+}
+
+type namespace struct {
+	file      *os.File
+	closed    bool
+	nsType    int
+	nsTypeStr string
+}
+
+// Returns an object representing the namespace referred to by @path
+func GetNS(nspath string) (Namespace, error) {
+	err := IsNSorErr(nspath)
+	if err != nil {
+		return nil, err
+	}
+
+	//cgroup  ipc  mnt  net  pid  user  uts
+	res := regexp.MustCompile(`^/proc/[^/]+/ns/([^/]+)`).FindStringSubmatch(nspath)
+	if len(res) != 2 {
+		return nil, fmt.Errorf("Error detecting namespace type from path: %s", nspath)
+	}
+
+	var nsType int
+	switch res[1] {
+	case "pid":
+		nsType = unix.CLONE_NEWPID
+	case "mnt":
+		nsType = unix.CLONE_NEWNS
+	case "net":
+		nsType = unix.CLONE_NEWNET
+	default:
+		return nil, fmt.Errorf("Error unsupported namespace requested: %s", res[1])
+	}
+
+	fd, err := os.Open(nspath)
+	if err != nil {
+		return nil, err
+	}
+	return &namespace{file: fd, nsType: nsType, nsTypeStr: res[1]}, nil
+}
+
+func (ns *namespace) Set() error {
+	if err := ns.errorIfClosed(); err != nil {
+		return err
+	}
+
+	if _, _, err := unix.Syscall(unix.SYS_SETNS, ns.Fd(), uintptr(ns.nsType), 0); err != 0 {
+		return fmt.Errorf("Error switching to ns %v: %v", ns.file.Name(), err)
+	}
+
+	return nil
+}
+
+func (ns *namespace) Close() error {
+	if err := ns.errorIfClosed(); err != nil {
+		return err
+	}
+
+	if err := ns.file.Close(); err != nil {
+		return fmt.Errorf("Failed to close %q: %v", ns.file.Name(), err)
+	}
+	ns.closed = true
+
+	return nil
+}
+
+func (ns *namespace) Path() string {
+	return ns.file.Name()
+}
+
+func (ns *namespace) Fd() uintptr {
+	return ns.file.Fd()
+}
+
+func (ns *namespace) errorIfClosed() error {
+	if ns.closed {
+		return fmt.Errorf("%q has already been closed", ns.file.Name())
+	}
+	return nil
+}
+
+func (ns *namespace) Do(toRun func(Namespace) error) error {
+	if err := ns.errorIfClosed(); err != nil {
+		return err
+	}
+
+	containedCall := func(hostNS Namespace) error {
+		threadNS, err := GetNS(ns.getCurrentThreadNSPath())
+		if err != nil {
+			return fmt.Errorf("failed to open current netns: %v", err)
+		}
+		defer threadNS.Close()
+
+		// switch to target namespace
+		if err = ns.Set(); err != nil {
+			return fmt.Errorf("error switching to ns %v: %v", ns.file.Name(), err)
+		}
+		defer threadNS.Set() // switch back
+
+		return toRun(hostNS)
+	}
+
+	// save a handle to current network namespace
+	hostNS, err := GetNS(ns.getCurrentThreadNSPath())
+	if err != nil {
+		return fmt.Errorf("Failed to open current namespace: %v", err)
+	}
+	defer hostNS.Close()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var innerError error
+	go func() {
+		defer wg.Done()
+		runtime.LockOSThread()
```

Comment:
I seem to remember even locking the thread posed issues with the namespace switching. We'll need to definitively prove this operation is safe. Worst case we can execute a binary here if we have to.

Reply:
Yes there were huge issues. Luckily with go 1.10 these issues are finally resolved and for as long as we start no go-routines in the callback we are finally save. Prior to 1.10 we would have to even start an extra binary, immediately lock there and then do the witch in that binary (cni plugins do it this way).

Reply:
Here a summary regarding to the current situation: https://www.weave.works/blog/linux-namespaces-golang-followup

Reply:
the "don't spawn goroutine" requirement here could get a little tricky depending on what logic we begin using. We'd have to inspect every suspect function call from external packages to ensure they don't execute a goroutine as well.

This isn't a big deal with the /dev/kvm/ device. Hopefully it's not something that impacts setup of other devices/resources

Reply:
Yes. In general  that is something all device-plugins which are written in go have to deal with too. Luckily the popular https://github.com/vishvananda/netlink library does not do that in any of its methods for setting up networking, since it is mostly used for such kind of container wiring operations which cross namespace boundaries.